File: src/pls/args.py
import argparse
import os
from pathlib import Path
from typing import Optional

from pls import __version__
from pls.enums.icon_type import IconType
from pls.enums.sort_order import SortOrder
from pls.enums.unit_system import UnitSystem
from pls.exceptions import ExecException


parser = argparse.ArgumentParser(
    prog="pls",
    description=(
        "`pls` is a better `ls` for developers. "
        "See https://github.com/dhruvkb/pls for more information."
    ),
)

########################
# Positional arguments #
########################


def directory(path_str: str) -> Path:
    """
    Parse the given path into a ``Path`` instance. The path is considered valid
    if it points to an existing directory.

    :param path_str: the path supplied as a CLI argument
    :return: the ``Path`` instance wrapping the supplied path
    :raise: ``ExecException``, if the path is invalid
    """

    path = Path(path_str).resolve()
    if not os.path.isdir(path):
        raise ExecException("`directory` must be a path to a directory")
    return path


parser.add_argument(
    "directory",
    type=directory,
    nargs="?",  # makes the `directory` arg optional
    default=os.getcwd(),
    help="the directory whose contents are to be listed",
)

######################
# Optional arguments #
######################

parser.add_argument(
    *["-v", "--version"],
    action="version",
    version=f"%(prog)s {__version__}",
    help="show the version of the codebase",
)


def file(path_str: str) -> Optional[Path]:
    """
    Parse the given path into a ``Path`` instance. The path is considered valid
    if nothing exists there or if it points to a file.

    :param path_str: the path supplied as a CLI argument
    :return: the ``Path`` instance wrapping the supplied path if it is valid,
        ``None`` otherwise
    """

    path = Path(path_str).resolve()
    if not path.exists() or (path.exists() and path.is_file()):
        return path
    else:
        return None


parser.add_argument(
    *["-e", "--export"],
    type=file,
    default=None,
    help="the path to the file where to write the exported HTML",
)

parser.add_argument(
    *["-i", "--icon"],
    type=IconType,
    choices=list(IconType),
    default=IconType.NERD,
    help="the type of icons to show with the files",
)
parser.add_argument(
    *["-s", "--sort"],
    type=SortOrder,
    choices=list(SortOrder),
    default=SortOrder.ASC,
    help="the direction in which to sort the files and directories",
)
parser.add_argument(
    *["-d", "--depth"],
    type=int,
    default=4,
    help="the max depth upto which to look for a `.pls.yml` file",
)
parser.add_argument(
    *["-u", "--units"],
    type=UnitSystem,
    choices=list(UnitSystem),
    default=UnitSystem.BINARY,
    help="the units to use when listing the size of files",
)

parser.add_argument(
    "--details",
    action="store_true",
    help="show details such as permissions, owner and size",
)
parser.add_argument(
    "--all",
    action="store_true",
    help="show all files, including those that would otherwise be hidden",
)

parser.add_argument(
    "--no-dirs",
    action="store_true",
    help="hide directories in the output",
)
parser.add_argument(
    "--no-files",
    action="store_true",
    help="hide files in the output",
)
parser.add_argument(
    "--no-align",
    action="store_true",
    help="turn off character alignment for leading dots",
)
parser.add_argument(
    "--no-dirs-first",
    action="store_true",
    help="mix directories and files, sorting them together",
)


args = parser.parse_args()
"""the CLI arguments parsed by ``argparse``"""

__all__ = ["args"]

File: src/config/args.rs
use crate::enums::{DetailField, SortField, Typ, UnitSys};
use crate::fmt::render;
use clap::Parser;
use log::warn;
use regex::bytes::{Regex, RegexBuilder};
use regex::Error as RegexError;
#[cfg(test)]
use std::ffi::OsString;
use std::path::PathBuf;

/// Parse the given string into a [`Regex`] while turning off Unicode mode.
///
/// The default implementations of `Regex` from string-types are all
/// Unicode-aware, so the builder pattern from [`RegexBuilder`] must be used.
///
/// Since user's can input regexes like 'ab.d', which can match invalid UTF-8
/// without Unicode mode, we must use `Regex` and `RegexBuilder` from
/// [`regex::bytes`].
fn regex_parser(s: &str) -> Result<Regex, RegexError> {
	RegexBuilder::new(s).unicode(false).build()
}

/// Represents the command-line arguments to `pls`.
///
/// `pls` picks sane defaults for the CLI arguments. If you prefer different
/// defaults, users can alias `pls` with their preferred flags set.
///
/// Note that `pls` allows for deep customisation using `.pls.yml` files, which
/// is not represented here. Refer to [`Conf`](crate::config::Conf) for those.
#[derive(Parser, Debug)]
#[command(
    author,
    version,
    about = render("<red bold>`pls`</> is a prettier and powerful `ls` for the pros."),
	long_about = render([
		"<red bold>`pls`</> is a prettier and powerful `ls` for the pros.\n",
		"<bold>Read docs:</> https://dhruvkb.github.io/pls",
		"<bold>Get source:</> https://github.com/dhruvkb/pls",
		"<bold>Sponsor:</> https://github.com/sponsors/dhruvkb"
	].join("\n")),
)]
pub struct Args {
	/// the paths to list, each of which may be a file or directory
	#[clap(default_value = ".")]
	pub paths: Vec<PathBuf>,

	/// the data points to show about each node
	#[clap(
		help_heading = "Detail view",
		short,
		long = "det",
		default_value = "none",
		value_enum
	)]
	pub details: Vec<DetailField>,

	/// show headers above columnar data
	#[clap(help_heading = "Detail view", short = 'H', long, default_value = "true", action = clap::ArgAction::Set)]
	pub header: bool,

	/// the type of units to use for the node sizes
	#[clap(
		help_heading = "Detail view",
		short,
		long,
		default_value = "binary",
		value_enum
	)]
	pub unit: UnitSys,

	/// display node names in multiple columns
	#[clap(help_heading = "Grid view", short, long, default_value = "false", action = clap::ArgAction::Set)]
	pub grid: bool,

	/// display node names column-first
	#[clap(help_heading = "Grid view", short = 'D', long, default_value = "false", action = clap::ArgAction::Set)]
	pub down: bool,

	/// display icons next to node names
	#[clap(help_heading = "Presentation", short, long, default_value = "true", action = clap::ArgAction::Set)]
	pub icon: bool,

	/// display node type suffixes after the node name
	#[clap(help_heading = "Presentation", short = 'S', long, default_value = "true", action = clap::ArgAction::Set)]
	pub suffix: bool,

	/// show symlink targets
	#[clap(help_heading = "Presentation", short = 'l', long, default_value = "true", action = clap::ArgAction::Set)]
	pub sym: bool,

	/// align items accounting for leading dots
	#[clap(help_heading = "Presentation", short, long, default_value = "true", action = clap::ArgAction::Set)]
	pub align: bool,

	/// the set of node types to include in the output
	#[clap(
		help_heading = "Filtering",
		short = 't',
		long = "typ",
		default_value = "all",
		value_enum,
        value_names = ["TYPES"],
	)]
	pub typs: Vec<Typ>,

	/// the importance cutoff to dim or hide unimportant files
	#[clap(help_heading = "Filtering", short = 'I', long, default_value = "0")]
	pub imp: i8,

	/// the pattern of files to selectively hide from the output
	#[clap(help_heading = "Filtering", short, long, value_parser = regex_parser)]
	pub exclude: Option<Regex>,

	/// the pattern of files to exclusively show in the output
	#[clap(help_heading = "Filtering", short, long, value_parser = regex_parser)]
	pub only: Option<Regex>,

	/// the set of fields to sort by, trailing `_` reverses the direction
	#[clap(help_heading = "Sorting", short, long = "sort", default_values = ["cat", "cname"], value_enum)]
	pub sort_bases: Vec<SortField>,
}

impl Default for Args {
	/// Create a new instance of `Args` parsing real command-line arguments.
	fn default() -> Self {
		let mut args = Args::parse();
		args.post_process();
		args
	}
}

impl Args {
	/// Create a new instance of `Args` parsing the given arguments.
	#[cfg(test)]
	pub fn raw<I, T>(itr: I) -> Self
	where
		I: IntoIterator<Item = T>,
		T: Into<OsString> + Clone,
	{
		Args::parse_from(itr)
	}

	/// Clean the parsed arguments and log any warnings that are raised.
	///
	/// The output of this function is similar to the format used by
	/// [`Exc`](crate::exc::Exc) as they serve similar purposes.
	fn post_process(&mut self) {
		self.clean().iter().for_each(|warning| {
			warn!("{warning}");
		});
	}

	/// Clean the parsed arguments to resolve conflicting arguments.
	///
	/// `pls` is intentionally lax about conflicting arguments, and will attempt
	/// to resolve conflicts in a way that is least surprising to the user. So,
	/// while the clean function generates warnings, they are not surfaced to
	/// the user and are primarily used for debugging.
	fn clean(&mut self) -> Vec<&str> {
		let mut warnings = vec![];

		self.details = DetailField::clean(&self.details);
		self.sort_bases = SortField::clean(&self.sort_bases);
		self.typs = Typ::clean(&self.typs);

		if self.grid && self.is_detailed() {
			// Multi-column mode is disabled when detailed mode is enabled.
			warnings.push("Detailed view disabled grid view.");
			self.grid = false;
		}

		// Headers cannot be shown outside of detailed view.
		if self.grid && self.header {
			warnings.push("Grid view disabled column headers.");
			self.header = false;
		}
		if !self.is_detailed() && self.header {
			warnings.push("Lack of metadata disabled column headers.");
			self.header = false;
		}

		if self.grid && self.sym {
			// Symlink targets cannot be shown in grid view.
			warnings.push("Grid view disabled symlink targets.");
			self.sym = false;
		}

		warnings
	}

	/* Getters */
	/* ======= */

	/// Get whether to render the output in detailed view using a table.
	fn is_detailed(&self) -> bool {
		self.details.len() >= 2
	}
}

#[cfg(test)]
mod tests {
	use super::Args;

	macro_rules! make_warning_test {
        ($($name:ident: $argv:expr => $msg:expr,)*) => {
            $(
                #[test]
                fn $name() {
                    let mut args = Args::raw($argv);
                    let warnings = args.clean();
                    assert!(warnings.contains(&$msg));
                }
            )*
        }
    }

	make_warning_test!(
		test_details_multi_col: ["pls", "--det", "ino", "--grid", "true"] => "Detailed view disabled grid view.",
		test_multi_col_and_header: ["pls", "--grid", "true", "--header", "true"] => "Grid view disabled column headers.",
		test_multi_col_and_sym: ["pls", "--grid", "true", "--sym", "true"] => "Grid view disabled symlink targets.",
	);

	macro_rules! make_clean_test {
        ($($name:ident: $argv:expr => $key:ident, $val:expr,)*) => {
            $(
                #[test]
                fn $name() {
                    let mut args = Args::raw($argv);
                    args.clean();
                    assert_eq!(args.$key, $val);
                }
            )*
        }
    }

	make_clean_test!(
		test_details_beats_multi_col: ["pls", "--det", "ino", "--grid", "true"] => grid, false,

		// Symlink target is only shown in detailed view.
		test_default_sym: ["pls"] => sym, true,
		test_default_sym_when_detailed: ["pls", "--det", "ino"] => sym, true,
		test_default_sym_when_multi_col: ["pls", "--grid", "true"] => sym, false,
		test_multi_col_beats_sym: ["pls", "--grid", "true", "--sym", "true"] => sym, false,

		// Header is only shown when detailed view is enabled and there is at least one detail field.
		test_default_header: ["pls"] => header, false,
		test_default_header_when_detailed: ["pls", "--det", "ino"] => header, true,
		test_default_header_when_multi_col: ["pls", "--grid", "true"] => header, false,
		test_multi_col_beats_header: ["pls", "--grid", "true", "--header", "true"] => header, false,
	);
}
